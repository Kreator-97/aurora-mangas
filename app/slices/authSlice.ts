import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { User } from '../../interfaces'

interface State {
  isLogged: Boolean;
  user    : User | null;
}

const initialState: State = {
  isLogged: false,
  user    : null,
}

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    login: (state, action:PayloadAction<{user: User}>) => {
      const { user } = action.payload
      return {...state, user, isLogged: true}
    },
  }
})

export const { login } = authSlice.actions
export default authSlice.reducer
