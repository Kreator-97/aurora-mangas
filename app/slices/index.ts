export { default as uiSlice } from './uiSlice'
export { default as authSlice } from './authSlice'
export { default as shoppingCartSlice } from './shoppingCartSlice'
