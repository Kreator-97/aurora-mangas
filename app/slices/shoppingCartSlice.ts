import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { ShoppingCart, ShoppingItem } from '../../interfaces'

const initialState: ShoppingCart = {
  items: [],
  total: 0
}

export const shoppingCartSlice = createSlice({
  name: 'shoppingCart',
  initialState,
  reducers: {
    addItem: (state, action: PayloadAction<ShoppingItem>) => {
      const { product, amount } = action.payload

      const item = state.items.find( item => item.product.id === product.id )

      if( !item ) {
        // if item is not included in previous items, we push it
        state.items.push({ product, amount })
        state.total = state.total + (product.price * amount )
        return
      }
      
      item.amount++
      state.total = state.total + (product.price * amount )
    },
    removeItem: (state, action: PayloadAction<{ id: string}>) => {
      const { id } = action.payload
      const items = state.items.filter( item => item.product.id !== id )
      
      state.items = items
      state.total = items.reduce( (acc, item) => (item.product.price * item.amount) + acc, 0)
    },
    setAmount: (state, action: PayloadAction<{amount:number, id: string}>) => {
      const { id, amount } = action.payload
      const index = state.items.findIndex( item => item.product.id === id )
      console.log(action.payload)
      if( index !== -1 ) {
        state.items[index].amount = amount
        state.total = state.items.reduce( (acc, item) => (item.product.price * item.amount) + acc, 0)
      }
    },
    setShoppingCart: (state, action: PayloadAction<ShoppingCart>) => {
      state = action.payload
      return state
    }
  }
})

export const { addItem, removeItem, setAmount, setShoppingCart } = shoppingCartSlice.actions
export default shoppingCartSlice.reducer
