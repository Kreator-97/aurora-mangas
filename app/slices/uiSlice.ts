import { createSlice } from '@reduxjs/toolkit'

interface State {
  isSidebarOpen: boolean;
  isShoppingCartOpen: boolean;
}

const initialState: State = {
  isSidebarOpen: false,
  isShoppingCartOpen: false,
}

export const uiSlice = createSlice({
  name: 'ui',
  initialState,
  reducers: {
    openSidebar: (state) => {
      state.isSidebarOpen = true
    },
    closeSidebar: (state) => {
      state.isSidebarOpen = false
    },
    closeShoppingCart: (state) => {
      state.isShoppingCartOpen = false
    },
    openShoppingCart: (state) => {
      state.isShoppingCartOpen = true
    }
  }
})

export const { closeSidebar, openSidebar, closeShoppingCart, openShoppingCart } = uiSlice.actions

export default uiSlice.reducer
