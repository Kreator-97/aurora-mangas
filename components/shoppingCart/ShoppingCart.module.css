
.cartlist-items {
  --footerHeight: 108px;
  --headerHeight: 65px;
  max-height: calc(100vh - ((var(--footerHeight)) + var(--headerHeight)) );
}

@media only screen and (max-device-width: 480px){
  .cartlist-items {
    max-height: calc(100vh - 4rem - ((var(--footerHeight)) + var(--headerHeight)) );
  }
 }