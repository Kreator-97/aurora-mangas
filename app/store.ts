import { configureStore } from '@reduxjs/toolkit'
import { uiSlice, authSlice, shoppingCartSlice } from './slices'

export const store = configureStore({
  reducer: {
    ui: uiSlice,
    auth: authSlice,
    cart: shoppingCartSlice
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
