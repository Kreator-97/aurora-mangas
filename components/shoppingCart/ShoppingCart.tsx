import { HiX } from 'react-icons/hi'
import Image from 'next/image'
import { useRouter } from 'next/router'

import { closeShoppingCart } from '../../app/slices/uiSlice'
import { dbLocal, formatPrice } from '../../util'
import { removeItem } from '../../app/slices/shoppingCartSlice'
import { useAppDispatch, useAppSelector } from '../../app/hooks'

import styles from './ShoppingCart.module.css'

const headerHeight = '65px'
const footerHeight = '108px'

export const ShoppingCart = () => {
  const dispatch = useAppDispatch()
  const { isShoppingCartOpen } = useAppSelector(state => state.ui)
  const { items, total } = useAppSelector(state => state.cart)
  const router = useRouter()

  const onClose = () => {
    dispatch( closeShoppingCart() )
  }

  const onRemove = (id:string) => {
    dispatch(removeItem({ id }))
    dbLocal.removeItemInLocal({items, total}, id)
  }

  const goToCheckout = () => {
    dispatch( closeShoppingCart() )
    router.push('/checkout')
  }

  return (
    <div
      className={`fixed top-0 right-0 min-h-screen h-screen min-w-[360px] bg-dark z-50 transition-transform ${isShoppingCartOpen ? 'translate-x-0' : 'translate-x-full'}`}>
      <div
        className='flex justify-between items-center border-b border-strokeLight p-4'
        style={{ height: headerHeight }}
      >
        <HiX
          size={28}
          onClick={ () => onClose()}
          cursor="pointer"
          data-testid="close-shopping-cart-icon"
        />

        <h2 className="text-center text-2xl">Carrito de compra</h2>
        <div className='w-7'></div>
      </div>
      <div
        style={{ paddingTop: '1rem', paddingBottom: '1rem' }}
        className={`flex flex-col gap-4 items-center overflow-y-scroll ${styles['cartlist-items']}`}
      >
        {
          items.map( item => {
            const title = `${item.product.serie.name} #${item.product.number}`
            return (
              <div key={item.product.id} className="flex">
                <Image
                  src={item.product.imgURL}
                  layout="fixed"
                  width={ 100 }
                  height={ 100 }
                  className="object-contain"
                  data-testid="cart-item"
                />
                <div className='px-2 w-52'>
                  <h2>{title}</h2>
                  <h2>Cantidad: {item.amount}</h2>
                  <h2>Subtotal: {formatPrice(item.amount * item.product.price)}</h2>
                  <button className='text-error text-sm' onClick={ () => onRemove(item.product.id) }>Eliminar del carrito</button>
                </div>
              </div>
            )
          })
        }
      </div>
      <div
        className='bg-dark absolute left-0 w-full px-4 py-4 border-t border-strokeLight'
        style={{ height: footerHeight }}
      >
        <h2 className='text-center text-2xl mb-2'>Total global: { formatPrice( total) }</h2>
        <button
          className='btn bg-accent w-full'
          onClick={ () => goToCheckout() }
        >Ir al resumen</button>
      </div>
    </div>
  )
} 
