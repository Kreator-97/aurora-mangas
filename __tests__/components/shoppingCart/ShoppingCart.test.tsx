import { fireEvent, render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'

import { mangas } from '../../fixtures/db'
import { store } from '../../../app/store'

const id = '001'

// mocks modules
import { removeItem } from '../../../app/slices/shoppingCartSlice'
import { closeShoppingCart } from '../../../app/slices/uiSlice'
import { ShoppingCart } from '../../../components/shoppingCart/ShoppingCart'
import { useRouter } from 'next/router'

jest.mock('../../../app/slices/uiSlice', () => {
  return {
    closeShoppingCart: jest.fn().mockReturnValue({payload: undefined, type: 'ui/closeShoppingCart'}),
  }
})

jest.mock('../../../app/slices/shoppingCartSlice', () => {
  return {
    removeItem: jest.fn().mockReturnValue({type: 'shoppingCart/removeItem', payload: { id: '001' }})
  }
})

jest.mock('../../../app/hooks', () => {
  const original = jest.requireActual('../../../app/hooks')
  return {
    ...original,
    useAppSelector: ((cart:any) => ({
      cart: { items: [{amount: 1, product: {...mangas[0], id}}], total: 0}}).cart),
  }
})

const router = { push: jest.fn() }
jest.mock('next/router', () => ({
  useRouter: () => (router)
}))

describe('test on shoppingCart component', () => {
  test('should to match with snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <ShoppingCart />
      </Provider>
    )
    expect(container).toMatchSnapshot()
    expect(screen.getByText('Carrito de compra')).toBeInTheDocument
    expect(screen.getByText(/Total global/)).toBeInTheDocument
    expect(screen.getByText('Ir al resumen')).toBeInTheDocument
  })

  test('onRemove function should to be called', () => {
    render(
      <Provider store={store}>
        <ShoppingCart />
      </Provider>
    )
    
    const closeIcon = screen.getByTestId('close-shopping-cart-icon')
    fireEvent.click(closeIcon)
    expect(closeShoppingCart).toHaveBeenCalled()
  })

  test('onRemove function should to be called with id to remove', () => {
    render(
      <Provider store={store}>
        <ShoppingCart />
      </Provider>
    )

    const btnRemove = screen.getByText('Eliminar del carrito')
    fireEvent.click(btnRemove)
    
    expect( removeItem ).toHaveBeenCalledWith({id})
  })

  test('should to go checkout when click on btn "Ir al resumen"', () => {
    render(
      <Provider store={store}>
        <ShoppingCart />
      </Provider>
    )

    const btnCheckout = screen.getByText('Ir al resumen')
    fireEvent.click(btnCheckout)
    expect(useRouter().push).toHaveBeenCalledWith('/checkout')
  })
})
