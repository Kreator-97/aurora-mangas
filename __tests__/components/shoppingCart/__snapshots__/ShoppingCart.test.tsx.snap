// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on shoppingCart component should to match with snapshot 1`] = `
<div>
  <div
    class="fixed top-0 right-0 min-h-screen h-screen min-w-[360px] bg-dark z-50 transition-transform translate-x-full"
  >
    <div
      class="flex justify-between items-center border-b border-strokeLight p-4"
      style="height: 65px;"
    >
      <svg
        cursor="pointer"
        data-testid="close-shopping-cart-icon"
        fill="currentColor"
        height="28"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
      <h2
        class="text-center text-2xl"
      >
        Carrito de compra
      </h2>
      <div
        class="w-7"
      />
    </div>
    <div
      class="flex flex-col gap-4 items-center overflow-y-scroll cartlist-items"
      style="padding-top: 1rem; padding-bottom: 1rem;"
    >
      <div
        class="flex"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 100px; height: 100px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
        >
          <img
            alt="One punch man no.12"
            class="object-contain"
            data-nimg="fixed"
            data-testid="cart-item"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
        <div
          class="px-2 w-52"
        >
          <h2>
            One Punch Man #12
          </h2>
          <h2>
            Cantidad: 
            1
          </h2>
          <h2>
            Subtotal: 
            $99.00
          </h2>
          <button
            class="text-error text-sm"
          >
            Eliminar del carrito
          </button>
        </div>
      </div>
    </div>
    <div
      class="bg-dark absolute left-0 w-full px-4 py-4 border-t border-strokeLight"
      style="height: 108px;"
    >
      <h2
        class="text-center text-2xl mb-2"
      >
        Total global: 
        $0.00
      </h2>
      <button
        class="btn bg-accent w-full"
      >
        Ir al resumen
      </button>
    </div>
  </div>
</div>
`;
